spec/lib/discourse_auth_cookie_spec.rb

```diff
@@ -0,0 +1,246 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+describe DiscourseAuthCookie do
+  fab!(:user) { Fabricate(:user) }
+
+  describe ".parse" do
+    it "does not attempt decryption or validations for v0 cookie" do
+      token = SecureRandom.hex
+      cookie = DiscourseAuthCookie.parse(token)
+      expect(cookie.token).to eq(token)
+    end
+
+    it "decrypts/validates signature for cookies that are < 32 chars" do
+      token = "asdad"
+      expect {
+        DiscourseAuthCookie.parse(token)
+      }.to raise_error(DiscourseAuthCookie::InvalidCookie)
+    end
+
+    it "decrypts/validates signature for v1 cookies" do
+      cookie = DiscourseAuthCookie.new(
+        token: SecureRandom.hex,
+        user_id: user.id,
+        trust_level: user.trust_level,
+        timestamp: 4.hours.ago,
+        valid_for: 100.days
+      ).to_text.dup
+
+      swap1 = 0
+      swap2 = cookie.split("").find_index { |c| c != cookie[swap1] }
+      cookie[swap1], cookie[swap2] = cookie[swap2], cookie[swap1]
+
+      expect {
+        DiscourseAuthCookie.parse(cookie)
+      }.to raise_error(DiscourseAuthCookie::InvalidCookie)
+
+      token = SecureRandom.hex
+      cookie = DiscourseAuthCookie.new(
+        token: token,
+        user_id: user.id,
+        trust_level: user.trust_level,
+        timestamp: 4.hours.ago,
+        valid_for: 100.days
+      ).to_text
+
+      parsed = DiscourseAuthCookie.parse(cookie)
+      expect(parsed.token).to eq(token)
+    end
+
+    it "correctly parses cookie fields" do
+      token = SecureRandom.hex
+      timestamp = 4.hours.ago
+      valid_for = 100.days
+      cookie = DiscourseAuthCookie.new(
+        token: token,
+        user_id: user.id,
+        trust_level: user.trust_level,
+        timestamp: timestamp,
+        valid_for: valid_for
+      ).to_text
+
+      parsed = DiscourseAuthCookie.parse(cookie)
+      expect(parsed.token).to eq(token)
+      expect(parsed.user_id).to eq(user.id)
+      expect(parsed.trust_level).to eq(user.trust_level)
+      expect(parsed.timestamp).to eq(timestamp.to_i)
+      expect(parsed.valid_for).to eq(valid_for.to_i)
+    end
+
+    it "works when there are missing fields in the cookie" do
+      token = SecureRandom.hex
+      timestamp = 4.hours.ago
+      valid_for = 100.days
+      cookie = DiscourseAuthCookie.new(
+        token: token,
+        user_id: nil,
+        trust_level: nil,
+        timestamp: timestamp,
+        valid_for: valid_for
+      ).to_text
+
+      parsed = DiscourseAuthCookie.parse(cookie)
+      expect(parsed.token).to eq(token)
+      expect(parsed.user_id).to eq(nil)
+      expect(parsed.trust_level).to eq(nil)
+      expect(parsed.timestamp).to eq(timestamp.to_i)
+      expect(parsed.valid_for).to eq(valid_for.to_i)
+    end
+
+    it "does not require cookie fields to be in any particular order" do
```

Comment:
Hmm personally I don't find this test very important so I feel like we can just drop it.